In [1]:
# Imported all the dependencies and modules needed:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from imblearn.over_sampling import RandomOverSampler
from collections import Counter

---

## Spliting the Data into Training and Testing Sets:

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame:
file_path = Path('Resources/lending_data.csv')
lending_df = pd.read_csv(file_path)

# Reviewed the DataFrame:
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Created the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separated the data into labels and features:
# First, separated the 'y' variable, which was the labels:
y = lending_df['loan_status']

# Then, separated the 'X' variable, which was the features:
X = lending_df.drop(columns='loan_status')

In [4]:
# Reviewed the 'y' variable Series:
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [5]:
# Reviewed the 'X' variable DataFrame:
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Checked the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Checked the balance of the target values:
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Split the data using 'train_test_split', seshaped the 'y' variable to better fit the data into the function, and
# assigned a 'random_state' of 1:
y = y.values.reshape(-1,1)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Creating a Logistic Regression Model with the Original Data:

###  Step 1: Fitted a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Created a 'StandardScaler' instance, fitted it, and scaled the data to standardise the features, remove the mean,
# and scale to unit variance, so that the 'LogisticRegression' behaves better:
scaler = StandardScaler()
X_scaler = scaler.fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Instantiated the 'Logistic Regression' model, and assigned a 'random_state' parameter of 1 to the model:
classifier = LogisticRegression(random_state=1)

# Fitted the model using the training data and used the 'ravel()' function to fit the original array:
lr_model = classifier.fit(X_train_scaled, y_train.ravel())

### Step 2: Saved the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Made a prediction using the testing data:
testing_predictions = lr_model.predict(X_test_scaled)

### Step 3: Evaluated the model’s performance by doing the following:

* Calculated the accuracy score of the model,

* Generated a confusion matrix,

* Printed the classification report.

In [10]:
# Printed the 'balanced_accuracy' score of the model:
balanced_acc_score = balanced_accuracy_score(y_test, testing_predictions)
print(f"The accuracy score of the model is: {balanced_acc_score}")

The accuracy score of the model is: 0.9889115309798473


In [11]:
# Generated a confusion matrix for the model:
test_cm = confusion_matrix(y_test, testing_predictions)
test_cm_df = pd.DataFrame(test_cm, index=['Actual 0', 'Actual 1'], columns=['Predicted 0', 'Predicted 1'])
print('Confusion matrix:')
display(test_cm_df)

Confusion matrix:


Predicted 0  Predicted 1
Actual 0        18652          113
Actual 1           10          609

In [12]:
# Printed the classification report for the model:
testing_report = classification_report(y_test, testing_predictions)
print('Testing classification report:')
print(testing_report)

Testing classification report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.98      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answered the following question.

**Question:** How well does the logistic regression model predict both the <font color=green>`0` (healthy loan)</font> and <font color=darkred>`1` (high-risk loan)</font> labels?

**Answer:** **In summary**, the above logistic regression model works *very well* to classify both <font color=green>'healthy loans'</font> and <font color=darkred>'high-risk loans'</font>. The model is as accurate as it can be for classifying <font color=green>'healthy loans'</font> and very accurate for <font color=darkred>'high-risk loans'</font> —based on the F1 scores, which are <font color=green>100%</font> and <font color=darkred>91%</font> respectively. The classifications made by the model are very likely correct because the precision scores are very high —<font color=green>100%</font> for <font color=green>0's</font> and <font color=darkred>84%</font> for <font color=darkred>1's</font>—, and the model correctly finds nearly all the true <font color=darkred>'high-risk loans'</font> there are, as the recall scores were extremely high too —<font color=green>99%</font> for <font color=green>0's</font> and <font color=darkred>98%</font> for <font color=darkred>1's</font>.

The following is a **detailed** explanation of the preceding paragraph. The goal of the model is to classify the right loans into the right categories. Although the *accuracy* score seems very high, it does not truly depict accuracy because there is a high class imbalance between the <font color=green>0's</font> and <font color=darkred>1's</font> —only <font color=darkred>3.2%</font> of the target values were <font color=darkred>'high-risk loans'</font> whereas <font color=green>the rest</font> were <font color=green>'healthy loans'</font>. Therefore, that score cannot correctly distinguish between false <font color=green>'healthy loans'</font> and false <font color=darkred>'high-risk loans'</font>.

When there is such a class imbalance, it is better to look at the model's precision, recall, and F1 score. The *precision* score reveals how precise a model is. The score of <font color=green>'healthy loans'</font> is as high as it can be, which means that the model correctly predicts <font color=green>100%</font> of the <font color=green>'healthy loans'</font>. On the other hand, the precision score of <font color=darkred>'high-risk loans'</font> was <font color=darkred>84%</font>, which, although not perfect, is still a high score. Although being wrong 14% of the time can sound alarming, in our situation it is not because the costs of mis-classifying a <font color=green>'healthy loan'</font> as a <font color=darkred>'high-risk loan'</font> are not nearly as high as mis-classifying a <font color=darkred>'high-risk loan'</font> as a <font color=green>'healthy loan'</font>. The costs of the latter situation are better explained by the recall score.

The *recall* score calculates how many of the actual <font color=darkred>'high-risk loans'</font> the model capture as <font color=darkred>'high-risk loans'</font>. Hence, the higher the score, the lower the costs of mis-classifying <font color=darkred>this type</font> of loans. The recall score of <font color=darkred>'high-risk loans'</font> was <font color=darkred>98%</font>, which is extremely high. That is very desirable because the model correctly identified nearly all the <font color=darkred>'high-risk loans'</font>. Therefore, the probabilities of mis-classifying <font color=darkred>this type</font> of loans is extremely low.

Finally, the *F1 score* is the harmonic mean of precision and recall. It is an alternative to the arithmetic mean, and it is used when working with rates. The F1 score of <font color=green>'healthy loans'</font> is <font color=green>100%</font> —as high as it can be—, while the score of <font color=darkred>'high-risk loans'</font> is <font color=darkred>91%</font> —very high. Those F1 scores show that the model is very accurate, and thus, performs very well, even for <font color=darkred>'high-risk loans'</font> almost all of the time.

---

## Predicting a Logistic Regression Model with Resampled Training Data:

### Step 1: Used the `RandomOverSampler` module from the imbalanced-learn library to resample the data and make sure that the labels had an equal number of data points. 

In [13]:
# Instantiate the random oversampler model, and assigned a 'random_state' parameter of 1 to the model:
ros = RandomOverSampler(random_state=1)

# Fitted the original training data to the 'random_oversampler' model:
X_train, y_train = ros.fit_resample(X, y)

In [14]:
# Counted the distinct values of the resampled labels data:
print('Resampled dataset shape %s' % Counter(y_train))

Resampled dataset shape Counter({0: 75036, 1: 75036})


### Step 2: Used the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [15]:
# Created a 'StandardScaler' instance, fitted it, and scaled the data to standardise the features, remove the mean,
# and scale to unit variance, so that the 'LogisticRegression' behaves better:
scaler = StandardScaler()
X_scaler = scaler.fit(X_train)
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Instantiated the 'Logistic Regression' model, and assigned a 'random_state' parameter of 1 to the model:
classifier = LogisticRegression(random_state=1)

# Fitted the model using the resampled training data, and used the 'ravel()' function to fit the original array:
lr_model = classifier.fit(X_train_scaled, y_train.ravel())

# Made a prediction using the testing data:
testing_predictions = lr_model.predict(X_test_scaled)

### Step 3: Evaluated the model’s performance by doing the following:

* Calculated the accuracy score of the model,

* Generated a confusion matrix,

* Printed the classification report.

In [16]:
# Printed the 'balanced_accuracy' score of the model:
balanced_acc_score = balanced_accuracy_score(y_test, testing_predictions)
print(f"The accuracy score of the model is: {balanced_acc_score}")

The accuracy score of the model is: 0.9934916041318802


In [17]:
# Generated a confusion matrix for the model:
test_cm = confusion_matrix(y_test, testing_predictions)
test_cm_df = pd.DataFrame(test_cm, index=['Actual 0', 'Actual 1'], columns=['Predicted 0', 'Predicted 1'])
print('Confusion matrix:')
display(test_cm_df)

Confusion matrix:


Predicted 0  Predicted 1
Actual 0        18642          123
Actual 1            4          615

In [18]:
# Printed the classification report for the model:
testing_report = classification_report(y_test, testing_predictions)
print('Testing classification report:')
print(testing_report)

Testing classification report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.83      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** **In summary**, oversampling the <font color=darkred>'high-risk loans'</font> improved the performance of the previous logistic regression model, which makes this improved model the safest to make decisions on <font color=darkred>that type</font> of loans. In general, this model works *very well* to classify both <font color=green>'healthy loans'</font> and <font color=darkred>'high-risk loans'</font>, as it is 99% accurate. The classifications made by the model are very likely correct because the precision scores are very high —<font color=green>100%</font> for <font color=green>0's</font> and <font color=darkred>83%</font> for <font color=darkred>1's</font>—, and this model also correctly finds nearly all the true <font color=darkred>'high-risk loans'</font> there are, as the recall score was extremely high: 99%.

The following is a **detailed** explanation of the preceding paragraph. Due to the oversampling of <font color=darkred>'high-risk loans'</font>, both classifications are now balanced, so the *accuracy* score can be trusted. It is very high —99%—, which means that this improved model is extremely accurate.

This improved model's precision and recall scores can shed more light on how well it performs. The *precision* score of <font color=green>'healthy loans'</font> is still as high as it can be, which means that the model correctly predicts <font color=green>100%</font> of <font color=green>that type</font> of loans. However, the precision score of <font color=darkred>'high-risk loans'</font> decreased to <font color=darkred>83%</font>, which is still a high score. Again, the costs of mis-classifying a <font color=darkred>'high-risk loan'</font> as a <font color=green>'healthy loan'</font> are more important, so the recall score should have more weight.

The *recall* score of <font color=darkred>'high-risk loans'</font> in this improved model was <font color=darkred>99%</font>, which is 1% higher than the previous model's recall score. That makes this model ideal for correctly identifying <font color=darkred>'high-risk loans'</font> and reduce mis-classification costs as much as possible.